File: skills/playwright/src/commands/navigate.rs
use crate::browser::BrowserSession;

fn normalize_url(url: &str) -> String {
    if url.starts_with("http://") || url.starts_with("https://") {
        url.to_string()
    } else {
        format!("https://{url}")
    }
}

pub async fn goto(session: &mut BrowserSession, url: &str) -> anyhow::Result<serde_json::Value> {
    let url = normalize_url(url);
    session.page.goto(&url).await?;

    Ok(serde_json::json!({
        "url": session.url().await,
        "title": session.title().await,
    }))
}

pub async fn back(session: &mut BrowserSession) -> anyhow::Result<serde_json::Value> {
    session
        .page
        .evaluate("window.history.back()")
        .await?;
    tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;

    Ok(serde_json::json!({
        "url": session.url().await,
        "title": session.title().await,
    }))
}

pub async fn forward(session: &mut BrowserSession) -> anyhow::Result<serde_json::Value> {
    session
        .page
        .evaluate("window.history.forward()")
        .await?;
    tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;

    Ok(serde_json::json!({
        "url": session.url().await,
        "title": session.title().await,
    }))
}

pub async fn reload(session: &mut BrowserSession) -> anyhow::Result<serde_json::Value> {
    session.page.reload().await?;

    Ok(serde_json::json!({
        "url": session.url().await,
        "title": session.title().await,
    }))
}

File: skills/playwright/src/commands/navigate.ts
import { getBrowser } from "../browser.js";

export interface NavigateResult {
	success: boolean;
	url: string;
	title: string;
	timestamp: string;
	error?: string;
}

function result(
	success: boolean,
	url: string,
	title: string,
	error?: string,
): NavigateResult {
	return {
		success,
		url,
		title,
		timestamp: new Date().toISOString(),
		...(error && { error }),
	};
}

export async function navigate(url: string): Promise<NavigateResult> {
	try {
		const { page } = await getBrowser();

		// Add protocol if missing
		let targetUrl = url;
		if (!url.startsWith("http://") && !url.startsWith("https://")) {
			targetUrl = `https://${url}`;
		}

		await page.goto(targetUrl, { waitUntil: "domcontentloaded" });
		const title = await page.title();

		return result(true, page.url(), title);
	} catch (err) {
		const error = err instanceof Error ? err.message : String(err);
		return result(false, url, "", error);
	}
}

export async function back(): Promise<NavigateResult> {
	try {
		const { page } = await getBrowser();
		await page.goBack({ waitUntil: "domcontentloaded" });
		const title = await page.title();

		return result(true, page.url(), title);
	} catch (err) {
		const error = err instanceof Error ? err.message : String(err);
		return result(false, "", "", error);
	}
}

export async function forward(): Promise<NavigateResult> {
	try {
		const { page } = await getBrowser();
		await page.goForward({ waitUntil: "domcontentloaded" });
		const title = await page.title();

		return result(true, page.url(), title);
	} catch (err) {
		const error = err instanceof Error ? err.message : String(err);
		return result(false, "", "", error);
	}
}

export async function reload(): Promise<NavigateResult> {
	try {
		const { page } = await getBrowser();
		await page.reload({ waitUntil: "domcontentloaded" });
		const title = await page.title();

		return result(true, page.url(), title);
	} catch (err) {
		const error = err instanceof Error ? err.message : String(err);
		return result(false, "", "", error);
	}
}
